gen/__init__.py

```diff
@@ -492,38 +570,56 @@ def generate(
-        assert item['path'] != '/pkginfo.json'
-
-    # If there aren't any files for a cloud-config package don't make one start
-    # existing adding a pkginfo.json
-    if len(cc_package_files) > 0:
-        cc_package_files.append({
-            "path": "/pkginfo.json",
-            "content": "{}"})
-
-    for item in cc_package_files:
-        assert item['path'].startswith('/')
-        item['path'] = '/etc/mesosphere/setup-packages/dcos-provider-{}--setup'.format(
-            argument_dict['provider']) + item['path']
-        rendered_templates['cloud-config.yaml']['root'].append(item)
+    # Find all files which contain late bind variables and turn them into a "late bind package"
+    # TODO(cmaloney): check there are no late bound variables in cloud-config.yaml
+    late_files, regular_files = extract_files_containing_late_variables(
+        rendered_templates['dcos-config.yaml']['package'])
+    # put the regular files right back
+    rendered_templates['dcos-config.yaml'] = {'package': regular_files}
 
-    cluster_package_info = {}
+    def make_package_filename(package_id, extension):
+        return 'packages/{0}/{1}{2}'.format(
+            package_id.name,
+            repr(package_id),
+            extension)
 
     # Render all the cluster packages
+    cluster_package_info = {}
+
+    # Prepare late binding config, if any.
+    late_package = build_late_package(late_files, argument_dict['config_id'], argument_dict['provider'])
+    if late_variables:
+        # Render the late binding package. This package will be downloaded onto
+        # each cluster node during bootstrap and rendered into the final config
+        # using the values from the late config file.
+        late_package_id = PackageId(late_package['name'])
+        late_package_filename = make_package_filename(late_package_id, '.config')
+        write_yaml(late_package_filename, {'package': late_package['package']})
+        cluster_package_info[late_package_id.name] = {
+            'id': late_package['name'],
+            'filename': late_package_filename
+        }
+
+        # Add the late config file to cloud config. The expressions in
+        # late_variables will be resolved by the service handling the cloud
+        # config (e.g. Amazon CloudFormation). The rendered late config file
+        # on a cluster node's filesystem will contain the final values.
+        rendered_templates['cloud-config.yaml']['root'].append({
+            'path': '/etc/mesosphere/setup-flags/late-config.yaml',
+            'permissions': '0644',
+            'owner': 'root',
+            # TODO(cmaloney): don't prettyprint to save bytes.
+            # NOTE: Use yaml here simply to make avoiding painful escaping and
+            # unescaping easier.
+            'content': render_yaml({
+                'late_bound_package_id': late_package['name'],
+                'bound_values': late_variables
+            })})
+
+    # Render the rest of the packages.
```

Comment:
do we still need these .tar.xz packages? Would be really nice to just ship all packages as yaml files (And would significantly speed up `tox -e py35-unittests release/test_release.py::make_channel_artifacts` among others